File: src/com/easy/CanPlaceFlowers_605.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class CanPlaceFlowers_605 {
    // 2019年8月4日
    // Use simple one pass
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 1 ms, faster than 100.00%
    // Memory Usage: 39.3 MB, less than 94.80%
    public boolean canPlaceFlowers(int[] flowerbed, int n) {
        if (n <= 0) {
            return true;
        }
        int mCount = 0;
        for (int i = 0; i < flowerbed.length; i++) {
            if (flowerbed[i] == 0 && (i == 0 || flowerbed[i - 1] == 0)
                    && (i == flowerbed.length - 1 || flowerbed[i + 1] == 0)) {
                flowerbed[i] = 1;
                mCount++;
                if (mCount >= n) {
                    return true;
                }
            }
        }

        return false;
    }
}

File: Golang_src/easy/canPlaceFlowers_605.go
package easy

// Use greddy algorithm:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 16 ms, faster than 87.59%
// Memory Usage: 5.9 MB, less than 94.99%
func canPlaceFlowers(flowerbed []int, n int) bool {
	index, count := 0, 0
	for index < len(flowerbed) {
		if flowerbed[index] == 0 && (index == 0 || flowerbed[index-1] == 0) && (index == len(flowerbed)-1 || flowerbed[index+1] == 0) {
			flowerbed[index] = 1
			index++
			count++
		}
		if count >= n {
			return true
		}
		index++
	}
	return false
}
